feat: Implement SQL-based schema versioning and migration execution

gabrielmatau79 opened the pull request:
Resolve issue #65 Add SQL-based migration support for schema versioning

### What's included

- Add `buildPgDatabaseWithMigrations` with support for file-based .sql migrations
- Introduce schema_version table for version tracking
- Add buildPgDatabaseWithMigrations function with ordered execution logic into PostgresMessagePickupRepository
- Add setupMigrations.ts to scaffold initial migration files

lotharking commented on packages/postgres/src/utils/setupMigration.ts:
```diff
@@ -0,0 +1,125 @@
+  created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+);
+
+CREATE INDEX IF NOT EXISTS queuedmessage_connectionId_index ON queuedmessage (connectionId);
```
IMHO, ensuring that all tables have appropriate primary keys and indices for optimized query performance is essential, especially as the tables grow. For instance, adding an index on the created_at and state columns can significantly improve search performance

gabrielmatau79 replied:
You're absolutely right. The SQL scripts in the migration folder have the correct structure, ensuring what you mentioned. This script was created to generate the initial SQL migration files, and I'm not sure if it's worth keeping.

lotharking commented on packages/postgres/src/utils/setupMigration.ts:
```diff
@@ -0,0 +1,125 @@
+CREATE INDEX IF NOT EXISTS queued_message_connection_id_idx ON queued_message (connection_id);
+
+CREATE TABLE IF NOT EXISTS live_session (
+  session_id VARCHAR(255) PRIMARY KEY,
```
I prefer reviewing the use of `VARCHAR(255)` for primary keys. I think using `UUID` as the primary key type would be more consistent and reliable in terms of uniqueness across multiple systems and data sources.

gabrielmatau79 replied:
I'll update the schema to reflect this  thanks again for the suggestion!

lotharking commented on packages/postgres/src/utils/setupMigration.ts:
```diff
@@ -0,0 +1,125 @@
+      console.warn(`⚠️  Skipped existing file: ${file.name}`)
+    }
+  }
+}
```
Are you cleaning up the old tables after the data migration into the main tables?

gabrielmatau79 replied:
Yes, the SQL scripts should include the commands to clean up and drop the tables, just like migrations/002-schema-upgrade.sql, which is the latest database version used by this package.

lotharking commented on packages/postgres/src/utils/setupMigration.ts:
```diff
@@ -0,0 +1,125 @@
+DO $$
+BEGIN
+  IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'message_state') THEN
+    CREATE TYPE message_state AS ENUM ('pending', 'sending');
```
Why not create a new table to store the state? In the future, editing the state directly in the `queued_message` table could become confusing.

lotharking commented on packages/postgres/src/utils/setupMigration.ts:
```diff
@@ -0,0 +1,125 @@
+import fs from 'fs'
```
Why do you use TypeORM? This library has migration integrations, creates the migrations table, allows rollback, and offers other useful functions.

gabrielmatau79 replied:
This script was initially created to generate the first SQL migration files. It contains both the initial version of the Message Pickup Repository database and the latest version currently used by the package. The fundamental purpose was to preserve the database history.

However, to avoid confusion, I will remove this script. From now on, SQL files will be created directly with the corresponding migration queries.

lotharking commented on packages/postgres/src/utils/setupMigration.ts:
```diff
@@ -0,0 +1,125 @@
+import fs from 'fs'
+import path from 'path'
```
Are you adding tests for CI/CD process on Github?

gabrielmatau79 replied:
setupMigration has been removed to avoid confusion. The SQL scripts are now located in the migrations folder. the monorepo has a Continuous Integration

lotharking commented on packages/postgres/src/utils/buildPgDatabaseWithMigrations.ts:
```diff
@@ -0,0 +1,117 @@
+  targetDatabase: string,
+  migrationsDir: string = path.resolve(__dirname, '../migrations'),
+): Promise<void> {
+  const adminClient = new Client({
```
AdminClient and pool could be refactored

gabrielmatau79 replied:
Resolved!

lotharking commented on packages/postgres/src/utils/buildPgDatabaseWithMigrations.ts:
```diff
@@ -0,0 +1,117 @@
+    await client.query('COMMIT')
+  } catch (err) {
+    await client.query('ROLLBACK')
+    console.error('[migration] Migration process failed:', (err as Error).message)
```
Could add the log first to add clarity to the process

gabrielmatau79 replied:
Resolved!

gabrielmatau79 commented:
Thanks for the feedback! @lotharking  Based on the suggestions and review comments, the following changes have been applied:

- Logger integration into buildPgDatabaseWithMigrations.
- Refactored shared connection config into baseConfig to reduce duplication
- Changed column types in live_session session_id: VARCHAR(255) → UUID
- Changed column types in queued_message connection_id: VARCHAR(255) → UUID
- Improved migration performance with new indexes 
- Removed setupMigrations.ts script

lotharking commented:
i don't have any problem, it looks ok for me. What do you think @genaris ?

genaris reviewed (approved):
Well done. I think it is a good minimalistic way to support migrations.
